Modules/_threadmodule.c

```diff
@@ -810,17 +815,26 @@ local_clear(localobject *self)
     Py_CLEAR(self->dummies);
     Py_CLEAR(self->wr_callback);
     /* Remove all strong references to dummies from the thread states */
-    if (self->key
-        && (tstate = PyThreadState_Get())
-        && tstate->interp) {
-        for(tstate = PyInterpreterState_ThreadHead(tstate->interp);
```

Comment:
My own backport of this was keeping the code closer to the original:

```
 static int
 local_clear(localobject *self)
@@ -813,14 +818,20 @@
     if (self->key
         && (tstate = PyThreadState_Get())
         && tstate->interp) {
-        for(tstate = PyInterpreterState_ThreadHead(tstate->interp);
-            tstate;
-            tstate = PyThreadState_Next(tstate))
+        _PyRuntimeState *runtime = &_PyRuntime;
+        HEAD_LOCK(runtime);
+        tstate = PyInterpreterState_ThreadHead(tstate->interp);
+        HEAD_UNLOCK(runtime);
+        while (tstate) {
             if (tstate->dict && PyDict_GetItem(tstate->dict, self->key)) {
                 if (PyDict_DelItem(tstate->dict, self->key)) {
                     PyErr_Clear();
                 }
             }
+            HEAD_LOCK(runtime);
+            tstate = PyThreadState_Next(tstate);
+            HEAD_UNLOCK(runtime);
+        }
     }
     return 0;
 }
 ```

Reply:
That's fine but I would keep the implementation same across branches if possible.

Reply:
OK, if the new version behaves identically in 3.9, that's great! I just wasn't sure what the difference is.